### PR TITLE
fix Xrc721 interface bug

### PR DESCRIPTION
**Pull request by newyear6565**

### CODE REVIEW

**Comment by newyear6565** [comment]
TODO:  Synchronize xrc721 data and write to existing database

File: indexprotocol/actions/xrc20.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ const (
 	//6352211e -> ownerOf(uint256)
 	ownerOfString = "6352211e000000000000000000000000fea7d8ac16886585f1c232f13fefc3cfa26eb4cc"
 
+	// if for xrc721: if `from` is 0x0, this is a new mint token
+	zeroAddrString = "0000000000000000000000000000000000000000"
```
Comment (guo) [inline indexprotocol/actions/xrc20.go:58]: @CoderZhi is this true?

File: indexprotocol/actions/xrc20_test.go [hunk 2]
```diff
@@ -132,4 +135,55 @@ func TestXrc20(t *testing.T) {
 	require.Equal("100000", xrc20History[0].BlockHeight)
 	require.Equal("888", xrc20History[0].Index)
 	require.Equal("failure", xrc20History[0].Status)
+
+	// fro xrc 721
```
Comment (guo) [inline indexprotocol/actions/xrc20_test.go:139]: for

Comment (newyear6565) [reply indexprotocol/actions/xrc20_test.go]: fixed @guo

File: indexprotocol/actions/xrc20.go [hunk 3]
```diff
@@ -264,6 +275,16 @@ func (p *Protocol) checkTopics(topics, data string) bool {
 	return true
 }
 
+func (p *Protocol) checkXrc721Topics(topics, data string) bool {
+	if topics == "" || len(topics) != 256 || len(data) != 0 {
```
Comment (lzxm160) [inline indexprotocol/actions/xrc20.go:279]: why len(topics) have to be 256 and len(data) is 0?

File: indexprotocol/actions/xrc20.go (hunk 1 above)
Comment (CoderZhi) [reply indexprotocol/actions/xrc20.go:58]: yes, that's true

File: indexprotocol/actions/xrc20.go (hunk 3 above)
Comment (newyear6565) [reply indexprotocol/actions/xrc20.go:279]: in this website :http://erc721.org/,  event Transfer(address indexed _from, address indexed _to, uint256 indexed _tokenId);  4 param is indexed, so all data are in topics

File: indexprotocol/actions/xrc20.go [hunk 4]
```diff
@@ -210,8 +218,11 @@ func (p *Protocol) updateXrc20History(
 				continue
 			}
 			if isXrc721 {
-				xrc721HoldersStrs = append(xrc721HoldersStrs, "(?, ?, ?)")
-				xrc721HoldersArgs = append(xrc721HoldersArgs, l.Address, from, blk.Timestamp().Unix())
+				// for mint token, address `from` is empty
+				if strings.Compare(from, "") != 0 {
```
Comment (lzxm160) [inline indexprotocol/actions/xrc20.go:222]: is this really happend?i think address.FromBytes(0) returns special string, not ""

Comment (newyear6565) [reply indexprotocol/actions/xrc20.go:222]: > is this really happend?i think address.FromBytes(0) returns special string, not ""

i have changed `func ParseContractData` in xrc20.go.if the address is 0x0, return a "" string, so this address will be not saved into table `xrc721_holders`

when we query `tokenHolderAddresses` ,  0x0 address will not show

File: indexprotocol/actions/xrc20.go [hunk 5]
```diff
@@ -409,21 +430,28 @@ func ParseContractData(topics, data string) (from, to, amount string, err error)
...
+	var ioAddress address.Address
+	var ethAddress common.Address
 	fromEth := all[sha3Len+contractParamsLen-addressLen : sha3Len+contractParamsLen]
-	ethAddress := common.HexToAddress(fromEth)
-	ioAddress, err := address.FromBytes(ethAddress.Bytes())
-	if err != nil {
-		return
+	if strings.Compare(zeroAddrString, fromEth) != 0 {
+		ethAddress = common.HexToAddress(fromEth)
+		ioAddress, err = address.FromBytes(ethAddress.Bytes())
```
Comment (lzxm160) [inline indexprotocol/actions/xrc20.go:439]: I think this address do not need special handling,ioAddress will be io1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqd39ym7

Comment (newyear6565) [reply indexprotocol/actions/xrc20.go:439]: yes it's right , i just don't want to save datas about  0x0 address in table `xrc721_holders`

Comment (lzxm160) [reply indexprotocol/actions/xrc20.go:439]: I think we should keep data integrity in mysql and filter this address for frontend If the front end doesn't need this data

Comment (newyear6565) [reply indexprotocol/actions/xrc20.go:439]: ok i will revert it

**Pull request merged**